CMakeLists.txt

```diff
@@ -63,8 +63,16 @@ configure_file("${CMAKE_SOURCE_DIR}/cmake/templates/license.h.in" include/licens
 
 include(EthOptions)
 configure_project(TESTS)
-
-find_package(Z3 4.8.0)
+set(LATEST_Z3_VERSION "4.8.12")
+if (TESTS AND NOT EMSCRIPTEN)
+  find_package(Z3 ${LATEST_Z3_VERSION} REQUIRED)
```

Comment:
Won't `REQUIRED` cause an error if you don't have Z3 installed at all? We do want to support people without Z3. It's just that if it's installed, you need specifically 4.8.12.

Reply:
> It's just that older versions are acceptable for `solc` but not `soltest` and for such versions you should be able to build just `solc` successfully but building `soltest` should error out.

yeah, that's what you told me earlier.
Actually, I want to inform users to disable TESTS

Reply:
ok, I meant that building `soltest` should error out with older Z3. But without Z3 (or with `-DUSE_Z3=OFF`) it should still build.

Reply:
you only want to print a warning and build the tests even when we don't have latest z3. so I'll remove `REQUIRED`.

Reply:
Ah, I see where the problem is. I thought that with older Z3 installed cmake would report an error even without `REQUIRED` and that the only thing that `REQUIRE` changed was be the case where Z3 is not installed at all. But looks like I was wrong. It continues and just does not use the older version.

What I wanted to achieve is this:
- With `TEST=OFF`
    - Z3 missing -> continue without Z3
    - Z3 < 4.8.0 -> error (but looks like currently it continues without Z3 so it's fine to leave it at that)
- With `TEST=ON`
    - Z3 missing -> continue without Z3
    - Z3 < 4.8.12 -> error

Reply:
I think that in the `TEST=ON` + Z3 < 4.8.12 case printing a warning and continuing without Z3 would be fine. But we should not report an error just because Z3 is not installed. It's not required to run tests.

Reply:
with `REQUIRED` :
> * With `TEST=ON`
>   * Z3 < 4.8.12 -> error
without `REQUIRED` :
* With `TEST=ON`
   * Z3 < 4.8.12 -> continue without Z3 and print a warning

I like the `REQUIRED` version because it is more clear to me. but it is a problem because it will change the default behavior. b_win is an example. I guess we have a lot of users that build solidity without z3 and without `TEST=OFF`.

Reply:
So how about this: use `find_package()` with 4.8.0 as a requirement in all cases. But then use `if`s to check if "tests are enabled" + "Z3 was found"  + "Z3 version is not 4.8.12". If all of these are true, manually report an error.

I think this would handle missing Z3 well and still give us an error if Z3 is installed.

Reply:
good idea